### ISSUE TITLE
error C2679: binary '<<': no operator found which takes a right-hand operand of type 'std::string_view'

### ISSUE THREAD

**Issue by NancyLi1013** [simdjson/simdjson#1437]
Hi all,

I'm a member of Microsoft VCPKG team, when I updated `simdjson `to 0.8.1, it build failed with the following errors with Visual Studio 2017:

```
[1/3] C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x86\cl.exe   /TP -DSIMDJSON_IMPLEMENTATION_ARM64=0 -DSIMDJSON_THREADS_ENABLED=1 -Dsimdjson_EXPORTS -IF:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\src -IF:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\include /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /WX /W3 /sdl /w34714 -std:c++17 /showIncludes /Fosrc\CMakeFiles\simdjson.dir\simdjson.cpp.obj /Fdsrc\CMakeFiles\simdjson.dir\ /FS -c F:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\src\simdjson.cpp
FAILED: src/CMakeFiles/simdjson.dir/simdjson.cpp.obj 
C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x86\cl.exe   /TP -DSIMDJSON_IMPLEMENTATION_ARM64=0 -DSIMDJSON_THREADS_ENABLED=1 -Dsimdjson_EXPORTS -IF:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\src -IF:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\include /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /WX /W3 /sdl /w34714 -std:c++17 /showIncludes /Fosrc\CMakeFiles\simdjson.dir\simdjson.cpp.obj /Fdsrc\CMakeFiles\simdjson.dir\ /FS -c F:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\src\simdjson.cpp
The simdjson library is designed for 64-bit processors and it seems that you are not compiling for a known 64-bit platform. All fast kernels will be disabled and performance may be poor. Please use a 64-bit target such as x64, 64-bit ARM or 64-bit PPC.
f:\16105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\include\simdjson/dom/serialization.h(122): error C2679: binary '<<': no operator found which takes a right-hand operand of type 'std::string_view' (or there is no acceptable conversion)
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(508): note: could be 'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(std::basic_streambuf<char,std::char_traits<char>> *)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(480): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(const void *)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(460): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(long double)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(440): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(double)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(420): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(float)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(400): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(unsigned __int64)'
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(380): note: or       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(__int64)'
```

Seems it requires #include <string_view> in `include/simdjson/dom/serialization.h file`.

I have made a patch to fix this on vcpkg https://github.com/microsoft/vcpkg/pull/16105

Note: simdisjon can be built successfully with Visual Studio 2019.

Please help confirm if this is a bug on simdjson side.

Thanks in advance!

**Comment by lemire** [comment]
Please report the result of 
```
cmake --build build
```

or the equivalent. We do not support building directly from `cl.exe`. 

Though we already had Visual Studio 2017 tests, I added more just now, under GitHub Actions...

https://github.com/simdjson/simdjson/pull/1438/checks?check_run_id=1863643564

**Comment by lemire** [comment]
Please note also that we do not officially support 32-bit platforms such as Win32. We offer some kind of fallback support, but users who are limited to 32-bit systems should probably not use simdjson. It was not designed for legacy platforms.

**Comment by lemire** [comment]
I cannot reproduce the issue, please help us by providing more details.

**Comment by NancyLi1013** [comment]
Hi @lemire 

Thanks for looking into this issue.

I used Windows 10 x64 platform to build.

I tried to build as you proposed above `cmake --build`, the same error as before:
```
Checking Build System
    Building Custom Rule F:/20200105/vcpkg/buildtrees/simdjson/src/9f89b884fd-05b3113302/benchmark/checkperf-reference/master
  /jsonexamples/generated/CMakeLists.txt
    Building Custom Rule F:/20200105/vcpkg/buildtrees/simdjson/src/9f89b884fd-05b3113302/benchmark/checkperf-reference/master
  /src/CMakeLists.txt
cl : Command line warning D9002: ignoring unknown option '-pthread' [F:\20200105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b
3113302\benchmark\checkperf-reference\master\build\src\simdjson.vcxproj] [F:\20200105\vcpkg\buildtrees\simdjson\src\9f89b884f
d-05b3113302\benchmark\checkperf-parse.vcxproj]
    simdjson.cpp
    The simdjson library is designed for 64-bit processors and it seems that you are not compiling for a known 64-bit platfor
  m. All fast kernels will be disabled and performance may be poor. Please use a 64-bit target such as x64, 64-bit ARM or 64-
  bit PPC.
f:\20200105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\benchmark\checkperf-reference\master\include\simdjson/dom/ser
ialization.h(122): error C2679: binary '<<': no operator found which takes a right-hand operand of type 'std::string_view' (o
r there is no acceptable conversion) [F:\20200105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\benchmark\checkperf-ref
erence\master\build\src\simdjson.vcxproj] [F:\20200105\vcpkg\buildtrees\simdjson\src\9f89b884fd-05b3113302\benchmark\checkper
f-parse.vcxproj]
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(508): note: coul
  d be 'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(std::ba
  sic_streambuf<char,std::char_traits<char>> *)'
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(480): note: or
       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(const v
  oid *)'
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(460): note: or
       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(long do
  uble)'
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(440): note: or
       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(double)
  '
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(420): note: or
       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(float)'
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\ostream(400): note: or
       'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::operator <<(unsigne
  d __int64)'
```

**Comment by NancyLi1013** [comment]
Attach the build process here
[build.txt](https://github.com/simdjson/simdjson/files/5955846/build.txt)

**Comment by lemire** [comment]
@NancyLi1013 Thank you. I was able to reproduce the issue locally even if it does not get triggered in CI.

**Comment by lemire** [comment]
@NancyLi1013 

I will issue a patch shortly. I have identified the problem.

Thanks for the thorough report.

**Comment by lemire** [comment]
How it does not get triggered in CI is puzzling.

**Comment by lemire** [comment]
Fixed by Version 0.8.2 https://github.com/simdjson/simdjson/releases/tag/v0.8.2

Please note that you may need to specify a x64 target when compiling under Visual Studio 2017 even if you have a 64-bit Windows. If you are getting warnings during the build of this type...

> The simdjson library is designed for 64-bit processors and it seems that you are not compiling for a known 64-bit platform. All fast kernels will be disabled and performance may be poor.

Then your compiler is targeting x86 platforms *even* if you have a 64-bit OS. The simdjson library is designed for 64-bit systems.

If you have further problems, please open a new issue.

**Comment by NancyLi1013** [comment]
Thanks for your help and so quick fix @lemire.
